### DISCUSSION TITLE
intersection-env with ContinuousAction

### DISCUSSION THREAD

**Discussion by KohMat** [eleurent/highway-env#205] [Q&A] [answered]
Hello, recently I found your super good project. I would like to use intersection-env with ContinuousAction, but it looks like I need to change the source code. If you don't mind, could you help me to make the code work?

Here is the code I tried first.
```python
env = gym.make("intersection-v0")
env.configure(
    {
        "action": {
            "type": "ContinuousAction",
        },
    }
)
obs = env.reset()
done = False
while not done:
    action = [0.1, 0.5]
    obs, reward, done, info = env.step(action)

    env.render()
```
I ran this code and got the following error from the [`IntersectionEnv._make_vehicles`](https://github.com/eleurent/highway-env/blob/master/highway_env/envs/intersection_env.py#L159) function.
```
AttributeError: 'Vehicle' object has no attribute 'plan_route_to'
```
The error occurred because the [`plan_route_to`](https://github.com/eleurent/highway-env/blob/d6ef0d6d766c3b54a9f06cba3f09677d718109c3/highway_env/vehicle/controller.py#L62) function is a method of `ControlledVehicle` for `DiscreteMetaAction`, but is not a method of `Vehicle` and `BicycleVehicle` for `ContinuousAction`. To fix this error, I changed how to create controlled vehicles to the following to check the effect quickly, referring to `highway_env.create_vehicles` function. 
```python
def _make_vehicles(self, n_vehicles: int = 10) -> None:
  # Controlled vehicles
  self.controlled_vehicles = []
- for ego_id in range(0, self.config["controlled_vehicles"]):
-     ego_lane = self.road.network.get_lane(("o{}".format(ego_id % 4), "ir{}".format(ego_id % 4), 0))
-     destination = self.config["destination"] or "o" + str(self.np_random.randint(1, 4))
-     ego_vehicle = self.action_type.vehicle_class(
-                      self.road,
-                      ego_lane.position(60 + 5*self.np_random.randn(1), 0),
-                      speed=ego_lane.speed_limit,
-                      heading=ego_lane.heading_at(60)) \
-         .plan_route_to(destination)
-     ego_vehicle.SPEED_MIN = 0
-     ego_vehicle.SPEED_MAX = 9
-     ego_vehicle.SPEED_COUNT = 3
-     ego_vehicle.speed_index = ego_vehicle.speed_to_index(ego_lane.speed_limit)
-     ego_vehicle.target_speed = ego_vehicle.index_to_speed(ego_vehicle.speed_index)

-     self.road.vehicles.append(ego_vehicle)
-     self.controlled_vehicles.append(ego_vehicle)
-     for v in self.road.vehicles:  # Prevent early collisions
-         if v is not ego_vehicle and np.linalg.norm(v.position - ego_vehicle.position) < 20:
-             self.road.vehicles.remove(v)

+ for i in range(self.config["controlled_vehicles"]):
+     vehicle = self.action_type.vehicle_class(self.road, [i*20, 0], 2*np.pi*self.np_random.rand(), 0)
+     self.road.vehicles.append(vehicle)
+     self.controlled_vehicles.append(vehicle)
```
Then I got the 2nd error from `RegulatedRoad.is_conflict_possible`.
```
AttributeError: 'Vehicle' object has no attribute 'predict_trajectory_constant_speed'
```
The [`predict_trajectory_constant_speed`](https://github.com/eleurent/highway-env/blob/d6ef0d6d766c3b54a9f06cba3f09677d718109c3/highway_env/vehicle/controller.py#L190) is also a method of `ControlledVehicle`, and predicts the future position of the vehicle along the route calculated by `plan_route_to`.

Is it reasonable to change the `Vehicle` or `BicycleVehicle` to have the `plan_route_to` and `_predict_trajectory_constant_speed` functions?
Or how about making a ControlledVehicle's inheritance which has the act function overridden to get the manual control signal, and using it to instantiate the `controlled_vehicles`?
Or are there any ideas or solutions?

**Comment by eleurent** [comment, marked as answer]
Hi @KohMat,
Yeah sorry about that, it is a bit messy. The continuous action support came after this environment, and compatibility was not properly checked for all environments.

Here is what I think is should work best:

1. wrap the `plan_route_to()` call in a try/except block, so that it fails silently whenever it is applied to a vehicle class that does not implement route tracking. This makes sense in your case of continuous actions: you are directly controlling the vehicle's acceleration and steering, so there is not planned route for the vehicle to follow, it is up to the agent to achieve this through its policy.
2. `_predict_trajectory_constant_speed` is used by other vehicles in the scene to reason about the future trajectories of nearby vehicles, and take yield decisions in case of imminent collision. If the agent is controlling a Vehicle manually, with no planned route available, we can take another (stronger) assumption that the vehicle will keep driving at constant speed & constant steering in the next timesteps, or alternatively constant speed & constant direction (steering angle = 0). Thus, you should implement one of these two versions and add it to the Vehicle class.

You also suggested the possibility of using `ControlledVehicle` for the ego-vehicle, and override its `act()` method to handle continuous actions. I agree that this would work, in that the two missing methods `plan_route_to` and `_predict_trajectory_constant_speed` would then already exist. However, I prefer having a simple Vehicle class to implement only kinematics, and then adding some additional features (lane keeping / velocity control / routes) in the `ControlledVehicle` class. Adding another layer on top of all that to go back to continuous actions seem backward to me, plus the methods `plan_route_to` and `_predict_trajectory_constant_speed` implemented in `ControlledVehicle` would assume wrongly that the vehicle will effectively follow the planned route, which it won't if the `act()` method is overriden, so this could cause other problems elsewhere.

I hope this makes sense.
Edouard

**Reply by KohMat** [reply]
Thank you. I added the `predict_trajectory_constant_speed` method in the `Vehicle` class. This is my [branch](https://github.com/KohMat/highway-env/commit/13a47822e97825a23b76c5d7a4e790820c162838). The prediction type can be selected in the constructor. Now the` intersection-v0` can be executed with the `ContinuousAction`. :+1: 

There was one issue related to the reward. The environment calculates the[ `high_speed_reward`](https://github.com/eleurent/highway-env/blob/master/highway_env/envs/intersection_env.py#L69) with the `speed_index` which is a member of the `ControlledVehicle`.
```python
reward = self.config["collision_reward"] * vehicle.crashed \
         + self.config["high_speed_reward"] * (vehicle.speed_index == vehicle.SPEED_COUNT - 1)
```

Since the `speed_index` is not in the `Vehicle` class, I used the [`high_speed_reward`](https://github.com/eleurent/highway-env/blob/master/highway_env/envs/highway_env.py#L87) from the highway-v0 when the vehicle type is `Vehicle`. 
```python
if isinstance(vehicle, ControlledVehicle):
    reward += self.config["high_speed_reward"] * (vehicle.speed_index == vehicle.SPEED_COUNT - 1)
else:
    scaled_speed = utils.lmap(self.vehicle.speed, self.config["reward_speed_range"], [0, 1])
    reward += self.config["high_speed_reward"] * np.clip(scaled_speed, 0, 1)
```
I added the `reward_speed_range` in the default_config of the ` intersection-v0`. Techically, the MIN, MAX, COUNTS([link](https://github.com/eleurent/highway-env/blob/master/highway_env/envs/intersection_env.py#L192)) can be calculated from the `reward_range`. Honestly I don't know how to calculate them nicely. I set MIN, MAX, COUNTS manually. Do you have any ideas?

**Reply by eleurent** [reply]
I don't really think that you need to keep separate terms in the reward function depending on whether the speed is controlled or not, what really matters is that high speeds are rewarded. 

So I would advise to only use the more generic formula, that maps the speed range to a reward range.

The index/min/max/count system is just a special case of defining a linear map, but only usable in the context of `ControlledVehicle `which have discrete tracked speeds, so it should be replaced.

**Reply by eleurent** [reply]
Oh right, the current reward function only rewards driving at full speed, it is not a linear reward. But that is not crucial I think, a linear map will do just fine.

**Reply by KohMat** [reply]
Thank you so much for your detailed answer. I removed the special reward.

**Reply by eleurent** [reply]
That's great, thanks! I filed a PR so that everybody can benefit from your work :)
